From 277f7255f9befb281cd60fbf6ce7d194446c98bc Mon Sep 17 00:00:00 2001
From: Jonas Kvinge <jonas@jkvinge.net>
Date: Wed, 18 May 2022 00:16:10 +0200
Subject: [PATCH] QAbstractItemModel: Fix Qt::TextAlignmentRole when metatype
 is uint

When combining text alignment flags, it no longer works since the
metatype for example QVariant(Qt::AlignRight | Qt::AlignVCenter) is
uint, not int.

Fixes: QTBUG-103576
Change-Id: If0291b99606787081c4bc26fd00431f8a17a61a2
Reviewed-by: Volker Hilsheimer <volker.hilsheimer@qt.io>
(cherry picked from commit 54d81d118997fc4c238b7266571d220d7a6065f1)
---
 src/corelib/itemmodels/qabstractitemmodel_p.h | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/src/corelib/itemmodels/qabstractitemmodel_p.h b/src/corelib/itemmodels/qabstractitemmodel_p.h
index edc5ad611691..564862d0edae 100644
--- a/src/corelib/itemmodels/qabstractitemmodel_p.h
+++ b/src/corelib/itemmodels/qabstractitemmodel_p.h
@@ -181,10 +181,12 @@ template <typename T>
 T legacyEnumValueFromModelData(const QVariant &data)
 {
     static_assert(std::is_enum_v<T>);
-    if (data.userType() == qMetaTypeId<T>())
+    if (data.userType() == qMetaTypeId<T>()) {
         return data.value<T>();
-    else if (data.userType() == qMetaTypeId<int>())
+    } else if (std::is_same_v<std::underlying_type_t<T>, int> ||
+               std::is_same_v<std::underlying_type_t<T>, uint>) {
         return T(data.toInt());
+    }
 
     return T();
 }
@@ -192,10 +194,12 @@ T legacyEnumValueFromModelData(const QVariant &data)
 template <typename T>
 T legacyFlagValueFromModelData(const QVariant &data)
 {
-    if (data.userType() == qMetaTypeId<T>())
+    if (data.userType() == qMetaTypeId<T>()) {
         return data.value<T>();
-    else if (data.userType() == qMetaTypeId<int>())
+    } else if (std::is_same_v<std::underlying_type_t<typename T::enum_type>, int> ||
+               std::is_same_v<std::underlying_type_t<typename T::enum_type>, uint>) {
         return T::fromInt(data.toInt());
+    }
 
     return T();
 }
-- 
2.16.3

